README.md
[![astropy](http://img.shields.io/badge/powered%20by-AstroPy-orange.svg?style=flat)](http://www.astropy.org/)
[![mesmer-master Actions Status](https://github.com/bthorne93/mesmer/workflows/mesmer-master/badge.svg)](https://github.com/bthorne93/mesmer/actions)

# mesmer

Package to calculate the log probability of foregrounds.

## Description

Given a dataset $\mathbf{d}$, this package calculates, and samples from, the maximum 
likelihood set of foreground parameters, $P(\mathbf{\theta}}\mathbf{d})$.

This approach has been used in these papers:

1. [Simulated forecasts for primordial B-mode searches in ground-based experiments](https://arxiv.org/abs/1608.00551)
2. [The Simons Observatory: Science goals and forecasts](https://arxiv.org/abs/1808.07445)
3. [Removal of Galactic foregrounds for the Simons Observatory primordial gravitational wave search](https://arxiv.org/abs/1905.08888)

## Useage

[An example note book](examples/example.ipynb) is in the [`examples`](examples) folder.

To add SEDs, edit [the SED functions file](mesmer/seds.py). The likelihood is defined in [the likelihood file](mesmer/likelihood.py).

# License

This project is Copyright (c) Ben Thorne and licensed under
the terms of the BSD 3-Clause license. This package is based upon
the [Astropy package template](https://github.com/astropy/package-template)
which is licensed under the BSD 3-clause license. See the licenses folder for
more information.


CHANGES.rst
0.0.4 (unreleased)
=====

0.0.3 (2020-09-01)
=====
- Python 3.8 now required.
- Reimplemented to use jax.numpy.
- Added testing for mesmer.seds and mesmer.likelihood.
- Added example notebook implementing ML and HMC.

0.0.2 (2020-07-31)
=====

- Moved to CI / CD with GitHub Actions.

0.0.1 (2020-07-29)
==================

- Development release

mesmer/tests/test_sed.py
def test_fmatrix():
    import numpy as np
    from mesmer.seds import FMatrix

    freqs = np.logspace(1, 3, 100)
    components = ["cmb", "syncpl", "dustmbb"]
    fmatrix = FMatrix(components)

    parameters = {
        "nu": freqs,
        "nu_ref_d": 353.0,
        "nu_ref_s": 23.0,
        "beta_d": 1.5,
        "beta_s": -3.0,
        "T_d": 20.0,
    }

    output1 = fmatrix(**parameters)
    assert output1.shape == (len(components), len(freqs))

    parameters.pop("nu")
    # Check that the same answer is achieved by calling FMatrix
    # in a different way.
    output2 = fmatrix(nu=freqs, **parameters)
    np.testing.assert_array_equal(output1, output2)
    return


def test_dustmbb():
    from mesmer.seds import dustmbb
    from jax import grad
    import numpy as np

    nu = 100.
    nu_ref_d = 353.
    beta_d = 1.5
    T_d = 20.

    # compare analytic gradient to jax gradient
    df_db = grad(dustmbb, argnums=2)
    df_db_0 = df_db(nu, nu_ref_d, beta_d, T_d)
    df_db_1 = dustmbb(nu, nu_ref_d, beta_d, T_d) * np.log(nu / nu_ref_d)
    np.testing.assert_almost_equal(df_db_0, df_db_1)

def test_syncpl():
    from mesmer.seds import syncpl
    from jax import grad
    import numpy as np
    nu = 100.
    nu_ref_s = 23.
    beta_s = -3.

    # compare analytic gradient to jax gradient
    df_db = grad(syncpl, 2)
    df_db_0 = df_db(nu, nu_ref_s, beta_s)
    df_db_1 = (nu / nu_ref_s)**beta_s * np.log(nu / nu_ref_s)
    np.testing.assert_almost_equal(df_db_0, df_db_1)

mesmer/tests/test_likelihood.py
import pytest
from itertools import product

@pytest.fixture(params=list(product([-3., -2.9], [1.5, 1.6])), autouse=True, scope='function')
def fake_data(request):
    import numpy as np
    beta_s, beta_d = request.param

    nfreq = 12
    sens = 2
    npix = 100

    dust_template = 100. * np.ones((1, 2, npix))
    freq_d = 353.
    T_d = 20.

    sync_template = 5 * np.ones((1, 2, npix))
    freq_s = 23.

    cmb_template = 10. * np.ones((1, 2, npix))

    freq_obs = np.array([23., 30., 40., 95., 100., 145., 220., 270., 300., 353., 545, 800])
    assert len(freq_obs) == nfreq
    sensitivities = np.array([sens] * nfreq)


    hnkt_ref = 0.0479924466 * freq_d / T_d
    dust_fn = lambda nu: (nu / freq_d) ** (beta_d + 1) * (np.exp(hnkt_ref) - 1) / (np.exp(0.0479924466 * nu / T_d) - 1)
    sync_fn = lambda nu: (nu / freq_s) ** (beta_s)
    cmb_fn = lambda nu: np.exp(0.0176086761 * nu) * (0.0176086761 * nu / (np.exp(0.0176086761 * nu) - 1)) ** 2

    dust_scaling = np.array(list(map(dust_fn, freq_obs)))
    dust_scaled = dust_scaling[:, None, None] * dust_template

    sync_scaling = np.array(list(map(sync_fn, freq_obs)))
    sync_scaled = sync_scaling[:, None, None] * sync_template

    cmb_scaling = np.array(list(map(cmb_fn, freq_obs)))
    cmb_scaled = cmb_scaling[:, None, None] * cmb_template

    data = dust_scaled + cmb_scaled + sync_scaled
    covariance=np.ones((nfreq, 2, npix)) * sensitivities[:, None, None] ** 2

    noise = np.random.randn(nfreq, 2, npix) * np.sqrt(covariance)
    data += noise

    templates=(dust_template, sync_template, cmb_template)
    scaled_maps = (dust_scaled, sync_scaled, cmb_scaled)
    return (freq_obs, request.param[0], request.param[1], templates, scaled_maps, data, covariance)


def test_max(fake_data):
    """ This test runs the simplest possible component separation case where we know the
    model exactly, and there is no spatial variation in either amplitudes or spectral 
    indices.

    The data is generated as a three component model: CMB, dust, synchrotron, all of which have
    spatially uniform emission and SEDs. The test is run with multiple values of the spectral
    parameters.

    The observed data has very low noise (RMS 0.2 in each pixel at each frequency) so the 
    model should be very well constrained.

    The data is generated at Nfreq frequencies. Nfreq >~ 10 should result in very good results,
    suitable for testing the code.
    """
    from mesmer import LogProb
    import numpy as np
    from scipy.optimize import minimize
    from jax import grad, jacfwd, jacrev

    frequencies, true_beta_s, true_beta_d, templates, scaled_maps, data, covariance = fake_data
    dust_template, sync_template, cmb_template = templates
    dust_scaled, sync_scaled, cmb_scaled = scaled_maps
    true_params = np.array([true_beta_d, true_beta_s])
    model = {
        'dustmbb': {
            'varied': {
                'beta_d': [1.5, 0.5]
            },
            'fixed': {
                'T_d': 20.,
                'nu_ref_d': 353.,
            }
        },
        'syncpl': {
            'varied': {
                'beta_s': [-3., 0.5]
            },
            'fixed': {
                'nu_ref_s': 23.,
            }
        },
        'cmb': {
            'varied': {
            },
            'fixed': {
            }
        },
    }
    logprob = LogProb(data=data, covariance=covariance, frequencies=frequencies, model=model)

    dlnP_dTheta = grad(logprob, 0)

    print('Log prob derivative Bd', dlnP_dTheta(true_params, True))
    print('Log prob derivative compared to logprob: ', dlnP_dTheta(true_params) / logprob(true_params))


    hessian = jacfwd(jacrev(logprob))
    hessian_eval = hessian(true_params, True)
    print("Hessian evaluated", hessian_eval)

    fish_est = np.linalg.inv(hessian_eval)
    print("Fishers estimates of uncertainty: ", np.sqrt(fish_est[0, 0]), np.sqrt(fish_est[1, 1]))

    marg_errors =  np.sqrt(np.array([fish_est[0, 0], fish_est[1, 1]]))

    res = minimize(logprob, [true_beta_d, true_beta_s], tol=1e-9, args=(True), method='Powell')
    dust = logprob.get_amplitude_expectation(res.x, component='dustmbb')
    cmb = logprob.get_amplitude_expectation(res.x, component='cmb')
    sync = logprob.get_amplitude_expectation(res.x, component='syncpl')
    print(f"Dust, input mean {np.mean(dust_template)} recovered mean {np.mean(dust)}")
    print(f"CMB, input_mean {np.mean(cmb_template)} recovered mean {np.mean(cmb)}")
    print(f"Sync, input mean {np.mean(sync_template)} recovered mean {np.mean(sync)}")
    print(f"True parameters {true_params}, recovered parameters {res.x}")
    print(f"Parameter error {(true_params - res.x) / marg_errors}-sigma")

    # Assert that the difference between true parameters and recovered is less than 
    # three times the calculated uncertainty.
    np.testing.assert_array_less(true_params - res.x, 5. * marg_errors)
    return

mesmer/tools.py
import numpy as np

__all__ = ["WhiteNoise"]


class WhiteNoise(object):
    """ Simple white noise model. See eqs (1) and (2) from arxiv: 9705188.
    """

    def __init__(self, weights=None, sens=None, pol=True):
        if weights is not None:
            assert weights.ndim == 1
            self.nfreq = weights.shape[0]
            self.weights = weights

        if sens is not None:
            assert sens.ndim == 1
            self.nfreq = sens.shape[0]
            self.weights = self._sens_to_weight(sens)

        self.npol = 1
        if pol:
            self.npol = 2
        return

    def map(self, nside, seed=3232):
        """ Function to generate a realization of the noise level
        as a HEALPix map, with a given `nside`.

        Parameters
        ----------
        nside: int
            Nside parameter of the HEALPix map to be produced.
        seed: int (optional, default=3232)
            RNG seed.

        Returns
        -------
        ndarray
            Noise realization as a HEALPix map.
        """
        np.random.seed(seed)
        npix = 12 * nside ** 2
        pix_var = self._get_pix_var(npix)
        out = np.random.randn(self.nfreq, self.npol, npix)
        out *= np.sqrt(pix_var[:, None, None])
        return out

    def spectrum(self, lmax, fwhm=None):
        return self._weight_to_spec(lmax, fwhm)

    def get_pix_var_map(self, nside):
        """ Function to return a map containing the variance
        in each pixel at each frequency. This is a product
        consumed by the `hoover.likelihood.LogProb` object,
        and is provided for convenient use with that.

        Parameters
        ----------
        nside: int
            Nside parameter of the HEALPix maps to be produced.

        Returns
        -------
        ndarray
            Array containing the variance in each pixel at each frequency.
        """
        npix = 12 * nside ** 2
        pix_var = self._get_pix_var(npix)
        out = np.ones((self.nfreq, self.npol, npix))
        return out * pix_var[..., None, None]

    def _get_pix_var(self, npix):
        """ Pixel variance is given by:

        ..math::

            \\sigma_{\\rm pix}^2 = \\frac{w^{-1}}{4 \\pi } N_{\\rm pix}

        """
        return self.weights * npix / 4.0 / np.pi

    def _sens_to_weight(self, sens):
        return (
            4.0 * np.pi * sens ** 2 / (4.0 * np.pi * (180.0 / np.pi) ** 2 * 60.0 ** 2)
        )

    def _weight_to_spec(self, lmax, fwhm=None):
        """ Noise spectrum is given by:

        ..math:

            N_\\ell = w^{-1} B^2_{\\ell}

        where the beam is given by:

        ..math:

            B_\\ell = \\exp(-\\ell(\\ell + 1) \\theta_{\\rm FWHM}^2 / (8 \\log 2))

        """
        ells = np.arange(2, lmax + 1)
        pol_spec = _pol_spec(self.weights[:, None], ells, fwhm)
        nell = np.zeros((self.nfreq, 2, lmax + 1))
        nell[:, 0, 2:] = pol_spec
        return nell


def _pol_spec(weight, ells, fwhm=None):
    """ Function to calculate white noise power spectrum with option to
    add beam effects.

    Parameters
    ----------
    weight: ndarray
        Array containing weights of frequency channels, shape (Nfreq)
    ells: ndarray
        Array of multipoles over which to calculate the noise power spectrum.
    fwhm: float (optional, default=None)
        If not None, deconvolve a beam with this fwhm in radians.

    Returns
    -------
    ndarray
        Power spectrum of noise.
    """
    if fwhm is not None:
        print(fwhm)
        return weight / _gaussian_beam(fwhm, ells) ** 2
    else:
        return weight * np.ones_like(ells)


def _gaussian_beam(fwhm, ells):
    """ Function to calculate a Gaussian window function in harmonic space.

    Parameters
    ----------
    fwhm: float
        Full width at half-maximum of the Gaussian beam, in radians.
    ells: ndarray
        Array containing the multipoles over which to calculate the beam.

    Returns
    -------
    ndarray
        Array containing the beam window function.
    """
    return np.exp(-ells * (ells + 1) * fwhm ** 2 / (np.log(8) * 2))


mesmer/likelihood.py
""" This module contains the main likelihood object of the package.
:class:`mesmer.likelihood.LogProb`. This object is initialized by
providing a set of observed data and its covariance, and specifying
a model.
"""
import yaml
import h5py
import numpy as np
from .seds import FMatrix

__all__ = ["LogProb"]


class LogProb(object):
    """ Class calculating the marginalize posterior of spectral parameters
    given some observed data, its covariance, and a model for the emission
    of different components.
    This class links together the functions required to calculate the
    terms in Equation (A7) of 1608.00551.
    """

    def __init__(self, data=None, covariance=None, frequencies=None, model=None):
        """ The setup of this class requires the observed multifrequency
        sky maps, their pixel covariance (assumed diagonal in pixel space),
        the SED matrix (instance of `mesmer.FMatrix`), a dictionary
        containing the fixed parameter and their values, and a set of priors,
        which tell the likelihood which parameters are to be let vary.

        Parameters
        ----------
        data: ndarray
            Array of shape (Nfreq, Npol, Npix) containing the observed
            multi-frequency data.
        covariance: ndarray
            Array of shape (Nfreq, Npol, Npix) containing the pixel covariance
            of the array `data`.
        fmatrix: object, :class:`mesmer.FMatrix`
            Instance of :class:`mesmer.FMatrix` that defines the component
            scaling in the fitted sky model.
        fixed_parameters: dict
            Dictionary, where key, value pairs correspond to parameter names
            and values to be fixed.
        priors: dict
            Dictionary, where key, value pairs correpsond to which parameters
            are allowed to vary, and the corresponding mean and standard
            deviation of the Gaussian prior.
        """
        if (
            (data is not None)
            and (covariance is not None)
            and (frequencies is not None)
        ):
            self.data_setup(data, covariance, frequencies)

        if model is not None:
            self.model_setup(model)

    def __str__(self):
        """ Magic method for conveniently printing object summary.
        """
        msg = r"""
        Data
        ----
        Number of frequencies: {:d}
        Number of polarization channels: {:d}
        Number of pixels: {:d}
        Model
        -----
        Components: {:s}
        Free parameters: {:s}
        Fixed parameters: {:s}
        """.format(
            self.nfreq,
            self.npol,
            self.npix,
            " ".join(self._components),
            " ".join(self.free_parameters),
            " ".join(self._fixed_parameters),
        )
        return msg

    def __call__(self, theta, ret_neg=False):
        """ This method computes the log probability at a point in parameter
        space specified by `pars` and any additional `kwargs` that may
        overwrite members of the `pars` dictionary.

        Parameters
        ----------
        theta: ndarray
            Array containing the parameters that are being varied.
            These should match the parameters given as priors when
            instantiating this object.
        ret_neg: bool (optional, default=False)
            If True, return the negative log likelihood. Else return
            the positive.
        Returns
        -------
        float
            The log probability at this point in parameter space.
        """
        try:
            assert len(theta) == len(self.free_parameters)
        except AssertionError:
            raise AssertionError("Pass argument for each varied parameter.")
        lnprior = self._lnprior(theta)
        F = self._F(theta)
        N_T_inv = self._N_T_inv(theta, F=F)
        T_bar = self._T_bar(theta, F=F, N_T_inv=N_T_inv)
        lnP = _lnP(lnprior, T_bar, N_T_inv)
        # if ret_neg, return negative loglik1elihood. Convenient for use
        # with minimization functions to find ML.
        if ret_neg:
            return -lnP
        return lnP

    def data_setup(self, data, covariance, frequencies):
        self.nfreq, self.npol, self.npix = data.shape
        self.frequencies = frequencies
        self.N_inv_d = (data, covariance)

    def model_setup(self, model):
        """ Function to parse the model definition dictionary.

        The first level of the dictionary has keys corresponding
        to the different components, and must match one of the
        functions in `mesmer.seds`. The values of the keys are
        dictionaries with `fixed` and `varied` keywords. These
        contain further dictionaries with fixed parameter
        name / value pairs, and free parameter prior name / pairs.

        Parameters
        ----------
        model: dict
            Dictionary defining model.
        """
        self._components = list(model.keys())
        self._fmatrix = FMatrix(self._components)
        self._priors = {}
        self._fixed_parameters = {}
        for component in model.values():
            # get list of free parameters for this component
            prior = component.get("varied", None)
            # if there are not free parameters prior is None
            if prior is not None:
                self._priors.update(prior)
            # do the same for the fixed parameters
            fix_par = component.get("fixed", None)
            if fix_par is not None:
                self._fixed_parameters.update(fix_par)

        self.free_parameters = sorted(self._priors.keys())

    def load_data_from_hdf5(self, fpath, mc=0):
        """ Method to instantiate data attributes using data saved
        in an HDF5 archive.

        Parameters
        ----------
        fpath: str, Path
            Path to the hdf5 archive.
        record: str
            Record in which the data is stored. This could be the
            name of the group in which the datasets are saved.
        """
        with h5py.File(fpath, "r") as f:
            maps = f["maps"]
            data = maps["data_mc{:04d}".format(mc)][...]
            cov = maps["cov"][...]
            frequencies = maps.attrs["frequencies"]
        self.data_setup(data, cov, frequencies)

    @classmethod
    def load_data_from_hdf5_batch(cls, fpath, model=None):
        """ Method to instantiate data attrib utes using data saved
        in an HDF5 archive.
        Note that a simpler structure for this method would place
        the for loop within a single opening of the hdf5 file.
        However, this hogs the resource, and we want to be able
        to write to this file from other processes.

        Parameters
        ----------
        fpath: str, Path
            Path to the hdf5 archive.
        model: Path (optional, default=None)
            If specified, this is a path to a yaml file containing
            an SED model specification.

        Returns
        -------
        :class:`mesmer.likelihood.LogProb`
            Instantiated :class:`mesmer.likelihood.LogProb` object.
        """
        # get metadata from the simulation
        with h5py.File(fpath, "r") as f:
            maps = f["maps"]
            nmc = maps.attrs["monte_carlo"]
            frequencies = maps.attrs1["frequencies"]

        # loop over monte carlo realizations and return an initialized
        # `LogProb` object. Each of these objects will still need to
        # have a model loaded, unless the model keyword is defined.
        for i in range(nmc):
            with h5py.File(fpath, "r") as f:
                maps = f["maps"]
                cov = maps["cov"][...]
                data = maps["data_mc{:04d}".format(i)][...]
            if model is None:
                yield cls(data=data, covariance=cov, frequencies=frequencies)
            else:
                with open(model) as f:
                    cfg = yaml.load(f, Loader=yaml.FullLoader)
                yield cls(data=data, covariance=cov, frequencies=frequencies, model=cfg)

    @classmethod
    def load_model_from_yaml(cls, fpath):
        """ Method to load a model configuration from a ``yaml`` file.
        """
        with open(fpath) as f:
            model = yaml.load(f, Loader=yaml.FullLoader)
        return model["identifier"], cls(model=model["model"])

    def get_amplitude_expectation(self, theta, component=None):
        """ Convenience function to return the component-separated expected
        amplitudes, `T_bar`, taking care of the relevant reshaping.
        """
        T_bar = self._T_bar(theta)
        T_bar = np.moveaxis(T_bar.reshape(self.npol, self.npix, -1), 2, 0)
        if component is None:
            return T_bar
        assert component in self._components
        idx = self._components.index(component)
        return T_bar[idx]

    def get_amplitdue_covariance(self, theta, component=None):
        """ Convenience function to return the component covariances,
        `N_T_inv`, for a given set of spectral parameters.
        """
        N_T_inv = self._N_T_inv(theta)
        ncomp = len(self._components)
        N_T = np.linalg.inv(N_T_inv.reshape(self.npol, self.npix, ncomp, ncomp))
        if component is None:
            return N_T
        assert component in self._components
        idx = self._components.index(component)
        return N_T[:, :, idx, idx]

    @property
    def free_parameters(self):
        return self.__free_parameters

    @free_parameters.setter
    def free_parameters(self, val):
        self.__free_parameters = val

    @property
    def N_inv(self):
        return self.__N_inv

    @N_inv.setter
    def N_inv(self, val):
        self.__N_inv = val

    @property
    def N_inv_d(self):
        return self.__N_inv_d

    @N_inv_d.setter
    def N_inv_d(self, val):
        """ Setter method for the inverse-variance weighted data.

        Parameters
        ----------
        val: tuple(ndarray)
            Tuple containing the data and covariance arrays. Arrays must have
            the same shape, and are expected in shape (Nfreq, Npol, Npix).
        """
        (data, cov) = val
        try:
            assert data.ndim == 3
            assert cov.ndim == 3
        except AssertionError:
            raise AssertionError("Data must have three dimensions, Nfreq, Npol, Npix")
        shape = [self.npix * self.npol, self.nfreq]
        data = _reorder_reshape_inputs(data, shape)
        self.N_inv = 1.0 / _reorder_reshape_inputs(cov, shape)
        self.__N_inv_d = data * self.N_inv

    def theta_0(self, npoints=None, seed=7837):
        """ Function to generate a starting guess for optimization
        or sampling by drawing a random point from the prior.

        Parameters
        ----------
        npoints: int (optional, default=None)
            If npoints is not None, function returns an array
            of draws from the prior of dimension (npoints, ndim).
            This can be useful for initializing a set of optimization
            runs, or samplers.
        seed: int (optional, default=7837)
            Seed for the PRNG key used by `jax`. `jax` has a subtly
            different approach to random number generation to numpy.
            Worth reading about this before setting this number.

        Returns
        -------
        ndarray
            Array of length the number of free parameters.
        """
        if npoints is not None:
            shape = (npoints, len(self.free_parameters))
        else:
            shape = (len(self.free_parameters),)
        out = np.random.normal(*shape, dtype=np.float32)
        means = []
        stds = []
        for par in self.free_parameters:
            mean, std = self._priors[par]
            means.append(mean)
            stds.append(std)
        means = np.array(means)
        stds = np.array(stds)
        return means + out * stds

    def _F(self, theta):
        free_parameters = dict(zip(self.free_parameters, theta))
        params = {**free_parameters, **self._fixed_parameters, "nu": self.frequencies}
        return self._fmatrix(**params)

    def _N_T_inv(self, theta, F=None):
        if F is None:
            F = self._F(theta)
        return _N_T_inv(F, self.N_inv)

    def _T_bar(self, theta, F=None, N_T_inv=None):
        if F is None:
            F = self._F(theta)
        if N_T_inv is None:
            N_T_inv = self._N_T_inv(theta, F=F)
        return _T_bar(F, N_T_inv, self.N_inv_d)

    def _lnprior(self, theta):
        logprior = 0
        for arg, par in zip(theta, self.free_parameters):
            mean, std = self._priors[par]
            logprior += _log_gaussian(arg, mean, std)
        return logprior


def _N_T_inv(F: np.ndarray, N_inv: np.ndarray) -> np.ndarray:
    """Function to calculate the inverse covariance of component
    amplitudes, `N_T_inv. This is an implementation of Equation
    (A4) in 1608.00551. See also Equation (A10) for interpretation.

    Parameters
    ----------
    F: ndarray
        SED matrix
    N_inv: ndarray
        Inverse noise covariance.

    Returns
    -------
    ndarray
        N_T_inv, the inverse covariance of component amplitude.
    """
    Fprod = F[:, None, :] * F[None, :, :]
    return np.sum(Fprod[None, :, :, :] * N_inv[:, None, None, :], axis=3)


def _T_bar(F: np.ndarray, N_T_inv: np.ndarray, N_inv_d: np.ndarray) -> np.ndarray:
    """Function to calculate the expected component amplitudes, `T_bar`.
    This is an implementation of Equation (A4) in 1608.00551. See also
    Equation (A10) for interpretation.

    Parameters
    ----------
    F: ndarray
        SED matrix
    N_T_inv: ndarray
        Inverse component covariance.
    N_inv_d: ndarray
        Inverse covariance-weighted data.

    Returns
    -------
    ndarray
        T_bar, the expected component amplitude.
    """
    y = np.sum(F[None, :, :] * N_inv_d[:, None, :], axis=2)
    return np.linalg.solve(N_T_inv, y)


def _lnP(lnprior, T_bar, N_T_inv):
    """ Function to calculate the posterior marginalized over
    amplitude parameters.
    This function calcualtes Equation (A7), with the inclusion of a
    Jeffrey's prior of 1608.00551:
    .. math::

        p(b|d) \propto \exp\left[\frac{1}{2}\bar{T}^T N_T^{-1} \bar{T} \right]p_p(b)

    Parameters
    ----------
    lnprior: float
        Prior evaluated at the given point in parameter space
    T_bar: ndarray
        Array containing the component amplitude means calculated at this
        point in parameter space.
    N_T_inv: ndarray
        Component amplitude covariance calculated at this point in
        parameter space.

    Returns
    -------
    float
        Log likelihood of this set of spectral parameters.
    """
    return lnprior + 0.5 * np.einsum("ij,ijk,ik->", T_bar, N_T_inv, T_bar)


def _log_gaussian(par, mean, std):
    """ Function used to calculate a Gaussian prior.
    """
    return -0.5 * ((par - mean) / std) ** 2


def _reorder_reshape_inputs(arr, shape):
    """ Function to reorder axes and reshape dimensions of input data.
    This takes input data, assumed to be of shape: (Nfreq, Npol, Npix)
    and converts to shape (Npix * Npol, Nfreq), which is easier to work
    with in the likelihood.

    Parameters
    ----------
    arr: ndarray
        Numpy array with three dimensions.

    Returns
    -------
    ndarray
        Numpy array with two dimensions.
    """
    return np.moveaxis(arr, (0, 1, 2), (2, 0, 1)).reshape(shape).astype(np.float32)


mesmer/seds.py
""" This module contains the SEDs of the different physical
components that are allowed in the likelihood. The list
of currently allowed SEDs is:

* :py:func:`cmb`
* :py:func:`syncpl`
* :py:func:`sync_curvedpl`
* :py:func:`dustmbb`

These are referred to by name when defining the :py:class:`FMatrix`
object, which then composes the model as a linear sum of the
various components.

Notes
-----

All the SEDs in this file are given 




.. plot::
    :include-source:

    import matplotlib.pyplot as plt
    import numpy as np
    from mesmer.seds import *

    freqs = np.logspace(1, 3)
    c = cmb(freqs)
    s = syncpl(freqs, 100., -3)
    d = dustmbb(freqs, 100., 1.5, 20.)
    sc = sync_curvedpl(freqs, 100, -3, 0.05)
    fmatrix = FMatrix(['cmb', 'syncpl', 'dustmbb'])
    params = {
        'nu': freqs,
        'nu_ref_s': 100.,
        'nu_ref_d': 100.,
        'beta_d': 1.5,
        'T_d': 20.,
        'beta_s': -3.
    }
    f = np.sum(fmatrix(**params), axis=0)
    fig, ax = plt.subplots(1, 1)
    ax.loglog(freqs, c, label='cmb')
    ax.loglog(freqs, s, label='syncpl')
    ax.loglog(freqs, d, label='dustmbb')
    ax.loglog(freqs, sc, label='sync_curvedpl')
    ax.loglog(freqs, f, label='fmatrix (c+s+d)')
    ax.set_xlabel('frequency (GHz)')
    ax.set_ylabel('f(nu)')
    ax.set_xlim(23., 400.)
    ax.set_ylim(1e-2, 1e2)
    ax.legend()
    plt.show()

"""

import jax.numpy as np

__all__ = ["FMatrix", "cmb", "syncpl", "sync_curvedpl", "dustmbb"]
 

class FMatrix(object):
    """ Class to construct the foreground mixing matrix.
    This class models foreground SEDs of the different
    components of the sky model. This is an implementation of
    Equation (10) in 1608.00551, for a single set of
    spectral parameters.

    Examples
    --------

    >>> from mesmer.seds import FMatrix
    >>> fmatrix = FMatrix(['cmb', 'syncpl', 'dustmbb'])
    >>> pars = {
    ...     'nu': np.array([1, 100, 1000]),
    ...     'nu_ref_d' : 100.,
    ...     'nu_ref_s' : 100.,
    ...     'beta_s' : -3.,
    ...     'beta_d' : 1.5,
    ...     'T_d' : 20.,
    ... }
    >>> print(fmatrix(**pars).shape)
    (3, 3)

    """

    def __init__(self, components):
        """ Once instantiated this function evaluates the component SEDs
        with which this class was instantiated.

        As it is used in a variety of contexts, the
        :class:`mesmer.seds.FMatrix` can be called in a variety of ways.
        It must ultimately be passed the arguments for all of the component
        functions. This is usually done through keyword arguments, however,
        the frequency can be passed as the only positional argument.

        Parameters
        ----------
        components: list(str)
            List of function names. These functions, evaluated
            at a list of frequencies, will give the mixing
            matrix, `F`.

        Returns
        -------
        ndarray
            Array of shape (ncomps, nfreqs), representing the component
            SEDs evaluated for the requested frequencies.
        """
        assert isinstance(components, list)
        # check that the list of components correspond to existing functions
        for component in components:
            assert component in ["dustmbb", "syncpl", "cmb", "sync_curvedpl"]
        self.components = components

    def __call__(self, *args, **parameters) -> np.ndarray:
        """


        """
        if parameters:
            self.parameters = parameters
        if args:
            self.parameters.update(*args)
        # evaluate each component function for the point in parameter space
        # specified by `parameters` N.B that each `comp_func` is passed all
        # the parameters - this requires that no two functions share argument
        # names.
        outputs = [
            globals()[comp_func](**self.parameters)[None]
            for comp_func in self.components
        ]
        return np.concatenate(list(outputs))


def cmb(nu: np.ndarray, *args, **kwargs) -> np.ndarray:
    """ Function to compute CMB SED, as a function of frequency.

    This function computes the scaling of the Brightness temperature
    as a function of frequency. See for example Equation (3) of
    Ichiki + 2014 ("A concise review of foreground emission").

    The definition of brightness temperature is:

    .. math::

        \Delta T_{\rm CMB} = \frac{(e^x-1)^2}{x^2e^x}T_B

    The scaling is therefore given by:

    .. math::

        f_{CMB}(\\nu) = e^x \\frac{x}{(e^x - 1)^2}

    Parameters
    ----------
    nu: float, or array_like(float)
        Frequency in GHz, :math:`\\nu`.

    Returns
    -------
    ndarray
        CMB brightness temperature sed  evaluated at frequency nu.
    """
    x = 0.0176086761 * nu
    ex = np.exp(x)
    sed = ex * (x / (ex - 1)) ** 2
    return sed


def syncpl(
    nu: np.ndarray,
    nu_ref_s: np.float32,
    beta_s: np.float32,
    *args,
    **kwargs
) -> np.ndarray:
    """ Function to compute synchrotron power law SED, given by:

    .. math::
        f_{\\rm sync}(\\nu) = \\left(\\nu / \\nu_s \\right)^{\\beta_s}

    Parameters
    ----------
    nu: float, or array_like(float)
        Frequency in GHz, :math:`\\nu`.
    nu_ref_s: float
        Reference frequency in GHz, :math:`\\nu_s`.
    beta_s: float
        Power law index in RJ units, :math:`\\beta_s`.

    Returns
    -------
    array_like(float)
        Synchroton SED relative to reference frequency.
    """
    return (nu / nu_ref_s) ** beta_s


def sync_curvedpl(
    nu: np.ndarray,
    nu_ref_s: np.float32,
    beta_s: np.float32,
    beta_c: np.float32,
    *args,
    **kwargs
) -> np.ndarray:
    """ Function to compute curved synchrotron power law SED,
    given by:

    .. math::

        f_{\\rm syncpl}(\\nu) = \\left(\\nu / \\nu_s \\right)^{\\beta_s +
        \\beta_c \\ln(\\nu / \\nu_s)}

    Parameters
    ----------
    nu: float, or array_like(float)
        Frequency in GHz, :math:`\\nu`.
    nu_ref_s: float
        Reference frequency in GHz, :math:`\\nu_s`.
    beta_s: float
        Power law index in RJ units, :math:`\\beta_s``.
    beta_c: float
        Power law index curvature, :math:`\\beta_c`.

    Returns
    -------
    array_like(float)
        Synchroton SED relative to reference frequency.
    """
    return (nu / nu_ref_s) ** (beta_s + beta_c * np.log(nu / nu_ref_s))


def dustmbb(
    nu: np.ndarray,
    nu_ref_d: np.float32,
    beta_d: np.float32,
    T_d: np.float32,
    *args,
    **kwargs
) -> np.ndarray:
    """ Function to compute modified blackbody dust SED, given by:

    .. math::

        f_{\\rm dust}(\\nu) = \\left(\\nu / \\nu_d \\right)^{1 + \\beta_d}
        \\frac{e^{h \\nu_d / (k_B T_d)} - 1} {e^{h \\nu / (k_B T_d)} - 1}

    Recall that:

    .. math::

        B_\\nu(\\nu, T) = \\frac{2h\\nu^3}{c^2}\\frac{1}{e^{h\\nu/kT} - 1}.


    Parameters
    ----------
    nu: float or array_like(float)
        Freuency at which to calculate SED, :math:`\\nu`.
    nu_ref_d: float
        Reference frequency in GHz, :math:`\\nu_d`.
    beta_d: float
        Power law index of dust opacity, :math:`\\beta_d`.
    T_d: float
        Temperature of the dust, :math:`T_d`.

    Returns
    -------
    array_like(float)
        SED of dust modified black body relative to reference frequency.
    """
    x_to = np.float32(0.0479924466) * nu / T_d
    x_from = np.float32(0.0479924466) * nu_ref_d / T_d
    sed = (nu / nu_ref_d) ** (1.0 + beta_d)
    sed *= (np.exp(x_from) - 1) / (np.exp(x_to) - 1)
    return sed
